core/ir.ml

```diff
@@ -71,10 +71,11 @@ and binding =
 and special =
   | Wrong      of Types.datatype
   | Database   of value
-  | Lens       of value * Lens.Sort.t
-  | LensDrop   of value * string * string * value * Lens.Sort.t
-  | LensSelect of value * Lens.Phrase.t * Lens.Sort.t
-  | LensJoin   of value * value * string list * Lens.Phrase.t * Lens.Phrase.t * Lens.Sort.t
+  | Lens       of value * Lens.Type.t
+  | LensDrop   of value * string * string * value * Lens.Type.t
+  | LensSelect of value * [`Static of Lens.Phrase.t | `Dynamic of value] * Lens.Type.t
```

Comment:
Is there a reason these are polymorphic variants, rather than just using standard variants but namespaced properly?

Reply:
I think polymorphic variants is the right thing to use here, however as we have decided not to have them in the syntax tree or ir, you should probably use monomorphic variants.

Reply:
The terminology `Static`/`Dynamic` is slightly misleading here anyway.  The first case means that the convenient syntax for lens predicates is used, whereas the second case allows an arbitrary Links function instead (which allows constructing "dynamic" predicates).  I believe @rudihorn plans to get rid of the first case for the time being anyway, at least from the IR; the convenient syntax using `Lens.Phrase.t` can be desugared to the second case if needed.

Reply:
> 
> 
> The terminology `Static`/`Dynamic` is slightly misleading here anyway. The first case means that the convenient syntax for lens predicates is used, whereas the second case allows an arbitrary Links function instead (which allows constructing "dynamic" predicates). I believe @rudihorn plans to get rid of the first case for the time being anyway, at least from the IR; the convenient syntax using `Lens.Phrase.t` can be desugared to the second case if needed.

This is now changed. Static and Dynamic now both use the same syntax, but static predicates need to be inline functions that don't refer to any external variables (i.e. completely known statically). This allows a conversion from the sugar directly to `Lens.Phrase.t`, whereas dynamic lenses are first evaluated to IR and then converted to `Lens.Phrase.t`. I have only dropped the convenient syntax, but the static/dynamic differentiation still exists.

I used polymorphic variants, as this seems to be the way it is done for inline types in other projects, but I can convert it to a monomorphic variant.